rfc-146-production-deploy-access.md

```diff
@@ -0,0 +1,82 @@
+- we make ad hoc access exceptions for senior hires (such as lead technical architects) who need some access to systems, but don't necessarily need full production admin access
+- experience of doing 2ndline shifts is not necessarily the best preparation for working with production systems, particularly for frontend developers
+- our process for when junior developers and apprentices should be given production access isn't clear
+
+There have been numerous cases where teams with only a few people with production access are slowed down because deployments become a bottleneck.
+
+[RFC 128 - Continuous Deployment](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md) has in some ways made the situation worse. We wanted to make sure that rolling out continuous deployment didn't inadvertently weaken a policy, so we disabled merge access for people without production access on repositories which were continuously deployed.
+
+There were [attempts at formalising this process around 2016](https://docs.google.com/document/d/1lo1JAkFIeWrIl-7bzkzps8rYi2hscV6IOpDzc8dhFLs/edit#). Unfortunately we didn't go into detail on why 2ndline shifts were required. The theory is that the logic went something like this:
+
+- In the early days of GOV.UK our deployment pipeline and monitoring was immature
+- This meant that debugging issues with deployments often required ssh access
+- In turn, this meant that only people with ssh access were allowed to do deployments
+- Because it's easy to make mistakes when using highly privileged access through ssh, we required people to do a couple of 2ndline shifts to make sure they knew what they were doing
+
+Note that by this logic, the 2ndline requirements are defending against accidental mistakes, not against malicious access. We have separate defences against malicious employees (security clearance, probation etc.).
+
+## Proposal
+
+We should implement a new level of access - "Production Deploy Access". This level of access should include:
+
+- Permission to deploy apps in Jenkins (ideally without Jenkins admin permission)
+- Permission to merge continuously deployed applications
+- Readonly access to AWS, Fastly, logging systems, etc.
+- admin access to GOV.UK Signon (create and edit normal users)
+
+This level of access should not include:
+
+- ssh permission to production or staging
+- admin access to the Deploy Jenkins in production or staging
+- admin or poweruser access to AWS
+- admin access to other systems (e.g. Fastly etc.)
+- superadmin access to GOV.UK Signon (create and edit all user types and edit applications)
+
+Engineers should be granted this level of access once they've got the required level of security clearance and have been working on GOV.UK for at least one full month. Access should be granted at the discretion of the engineer's tech lead. Before approving access, tech leads should ensure that the engineer:
```

Comment:
Could/should this be revoked if the tech lead makes the call too early?Should it be a consideration? If so do we need to define the rules for it? Is there anything like that for current existing prod access?

Reply:
I echo Bill's point about the 1 month being arbitrary. Suggest we make this a guideline instead, e.g.:

```suggestion
Access should be granted at the discretion of the engineer's tech lead, once the engineer has the required level of security clearance. Before approving access, tech leads should ensure that the engineer:
```

Reply:
I don't think we've got a process for revoking access if it's granted too early - I think it should be rare enough that we can handle it exception by exception.